WebKit serialization tends to use copying when it should be moving

brentfulgham opened the pull request:
#### a83c1665d5a6a45ac4bdcb55eaadb38c0b6a3b75
<pre>
WebKit serialization tends to use copying when it should be moving
<a href="https://bugs.webkit.org/show_bug.cgi?id=291589">https://bugs.webkit.org/show_bug.cgi?id=291589</a>
<a href="https://rdar.apple.com/149327994">rdar://149327994</a>

Reviewed by Pascoe.

ResourceRequestBase, RequestData, and ResourceResponse Data are serialized through IPC using
move operations, but because these classes only offer copy-based constructors they are passing
through the copy code path instead of moving the memory. In the case of strings, URLs, Vectors
of Strings, Maps of Strings this is more costly than necessary.

This patch adds the relevant constructors and adopts them in more places.

* Source/WebCore/Modules/beacon/NavigatorBeacon.cpp:
(WebCore::NavigatorBeacon::sendBeacon): Use move operation.
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement): Ditto.
* Source/WebCore/loader/LinkLoader.cpp:
(WebCore::LinkLoader::preloadIfNeeded): Ditto.
* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::performDragOperation): Ditto.
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::setURL): Added.
* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData): Added.
(WebCore::ResourceRequestBase::ResourceRequestBase): Added.
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::ResourceResponseBase): Added.
(WebCore::ResourceResponseBase::setURL): Added.
* Source/WebCore/platform/network/ResourceResponseBase.h:
* Source/WebCore/platform/network/cf/ResourceRequest.h:
(WebCore::ResourceRequest::ResourceRequest): Add move constructor.
* Source/WebCore/platform/network/cf/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse): Ditto.
* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::fromResourceRequestData): Implemented move operations.
* Source/WebCore/platform/network/curl/ResourceRequest.h:
(WebCore::ResourceRequest::ResourceRequest): Add move constructor.
* Source/WebCore/platform/network/curl/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse): Ditto.
* Source/WebCore/platform/network/soup/ResourceRequest.h:
(WebCore::ResourceRequest::ResourceRequest): Ditto.
* Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp:
(WebCore::ResourceRequest::fromResourceRequestData): Implemented move operations.
* Source/WebCore/platform/network/soup/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse): Add move constructor.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadData): Use Move operators.
(WebKit::WebPage::loadAlternateHTML): Ditto.

Canonical link: <a href="https://commits.webkit.org/293818@main">https://commits.webkit.org/293818@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/894098d05c20341b3eea0cb3971474cdcba6b870

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99783 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19428 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9702 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104908 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50367 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101824 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19733 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27864 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75958 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33049 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102790 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15030 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90094 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56324 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14838 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8081 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49733 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84763 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8166 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107269 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26894 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19636 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84909 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27257 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86299 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84434 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21494 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29107 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6816 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20703 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26833 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32043 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26644 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29961 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28208 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebCore/platform/network/cf/ResourceRequest.h:
```diff
@@ -54,16 +54,32 @@ using ResourceRequestData = std::variant<ResourceRequestBase::RequestData, Resou
     }
 
-    ResourceRequest(const URL& url) 
+    ResourceRequest(URL&& url)
```
Should we add `explicit` here as well since this constructor takes in a single parameter?

brentfulgham replied:
We have a few places where a `ResourceRequest` is created on the fly from a URL when passing to a method. I tried making this (and the `const URL&` version) explicit, but it created various build failures that I don't think w want.

RupinMittal commented on Source/WebCore/platform/network/cf/ResourceRequest.h:
```diff
@@ -54,16 +54,32 @@ using ResourceRequestData = std::variant<ResourceRequestBase::RequestData, Resou
+    {
+    }
+
+    ResourceRequest(const URL& url)
```
Should we add `explicit` here as well since this constructor takes in a single parameter?

brentfulgham replied:
Ditto.

RupinMittal commented on Source/WebCore/platform/network/curl/ResourceRequest.h:
```diff
@@ -33,16 +33,32 @@ namespace WebCore {
     {
     }
 
+    ResourceRequest(URL&& url)
```
Should we use `explicit` here and on the below `ResourceRequest(const URL& url)` as well?

brentfulgham replied:
Ditto.

RupinMittal commented on Source/WebCore/platform/network/soup/ResourceRequest.h:
```diff
@@ -44,16 +44,32 @@ using ResourceRequestData = std::variant<ResourceRequestBase::RequestData, Resou
     {
     }
 
+    ResourceRequest(URL&& url)
```
Should we use `explicit` here and on `ResourceRequest(const URL& url)` below?

brentfulgham replied:
Ditto.

RupinMittal commented on Source/WebCore/platform/network/ResourceResponseBase.cpp:
```diff
@@ -60,6 +60,16 @@ ResourceResponseBase::ResourceResponseBase()
 {
 }
 
+ResourceResponseBase::ResourceResponseBase(URL&& url, String&& mimeType, long long expectedLength, const String& textEncodingName)
```
Should we use `String&& textEncodingName`?

brentfulgham replied:
In the callsites I found, there were cases where the text encoding was a const reference of something local, so we wouldn't be able to move it. So I left that argument alone.

We could add another constructor with a move version, which might be an improvement for the serializer (which moves all arguments). I could do a follow-up with that.

RupinMittal commented on Source/WebCore/platform/network/ResourceResponseBase.h:
```diff
@@ -232,6 +233,7 @@ class ResourceResponseBase {
     };
 
     WEBCORE_EXPORT ResourceResponseBase();
+    WEBCORE_EXPORT ResourceResponseBase(URL&&, String&& mimeType, long long expectedLength, const String& textEncodingName);
```
Should we use `String&& textEncodingName`?

brentfulgham replied:
Yes!

RupinMittal reviewed (commented):
Had a couple small questions :)

brentfulgham committed:
WebKit serialization tends to use copying when it should be moving
https://bugs.webkit.org/show_bug.cgi?id=291589
rdar://149327994

Reviewed by Pascoe.

ResourceRequestBase, RequestData, and ResourceResponse Data are serialized through IPC using
move operations, but because these classes only offer copy-based constructors they are passing
through the copy code path instead of moving the memory. In the case of strings, URLs, Vectors
of Strings, Maps of Strings this is more costly than necessary.

This patch adds the relevant constructors and adopts them in more places.

* Source/WebCore/Modules/beacon/NavigatorBeacon.cpp:
(WebCore::NavigatorBeacon::sendBeacon): Use move operation.
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement): Ditto.
* Source/WebCore/loader/LinkLoader.cpp:
(WebCore::LinkLoader::preloadIfNeeded): Ditto.
* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::performDragOperation): Ditto.
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::setURL): Added.
* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData): Added.
(WebCore::ResourceRequestBase::ResourceRequestBase): Added.
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::ResourceResponseBase): Added.
(WebCore::ResourceResponseBase::setURL): Added.
* Source/WebCore/platform/network/ResourceResponseBase.h:
* Source/WebCore/platform/network/cf/ResourceRequest.h:
(WebCore::ResourceRequest::ResourceRequest): Add move constructor.
* Source/WebCore/platform/network/cf/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse): Ditto.
* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::fromResourceRequestData): Implemented move operations.
* Source/WebCore/platform/network/curl/ResourceRequest.h:
(WebCore::ResourceRequest::ResourceRequest): Add move constructor.
* Source/WebCore/platform/network/curl/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse): Ditto.
* Source/WebCore/platform/network/soup/ResourceRequest.h:
(WebCore::ResourceRequest::ResourceRequest): Ditto.
* Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp:
(WebCore::ResourceRequest::fromResourceRequestData): Implemented move operations.
* Source/WebCore/platform/network/soup/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse): Add move constructor.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadData): Use Move operators.
(WebKit::WebPage::loadAlternateHTML): Ditto.

Canonical link: https://commits.webkit.org/293818@main
